# Exploring Hacker News Posts

Hacker News is a site created by the startup incubator Y Combinator for user-submitted stories. This stories receive votes and comments (similar to reddit). Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of the Hacker News listings can get hundreds of thousands of visitors as a result.

This report aims to explore, and analyze Hacker News dataset of about 20,000 rows to generate meaningful insight and make informed decision. Below are descriptions of the columns: 

id: the unique identifier from Hacker News for the post
- `title`: the title of the post
- `url`: the URL that the posts links to, if the post has a URL
- `num_points`: the number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
- `num_comments`: the number of comments on the post
- `author`: the username of the person who submitted the post
- `created_at`: the date and time of the post's submission


## Objectives

We're specifically interested in posts with titles that begin with either `Ask HN` or `Show HN`. Users submit `Ask HN` posts to ask the Hacker News community a specific question. Below are a few examples:

- Ask HN: How to improve my personal website?
- Ask HN: Am I the only one outraged by Twitter shutting down share counts?
- Ask HN: Aby recent changes to CSS that broke mobile?

Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just something interesting. Below are a few examples:

- Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform'
- Show HN: Something pointless I made
- Show HN: Shanhu.io, a programming playground powered by e8vm

We'll compare these two types of posts to determine the following:

- Do Ask HN or Show HN receive more comments on average?
- Do posts created at a certain time receive more comments on average?


In [2]:
from csv import reader 

# create open and read file into list of lists

openedFile = open('hacker_news.csv')
readFile = reader(openedFile)
hn = list(readFile)
headers = hn[0]
hn = hn[1:]

print(f'No of posts {len(hn)}\n')
print(headers )
print(hn[:5] )

No of posts 20100

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12'

## Extracting Ask HN and Show HN Posts

Above, we have explored the first 5 data (with header, for clarity). However, since we're only concerned with post titles beginning with Ask HN or Show HN, we'll create new lists of lists containing just the data for those titles.

In [3]:
ask_posts = []
show_posts = []
other_posts = []

# seperating hn list into sub categorized lists

for row in hn:
    title = row[1].lower()
    
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        

In [4]:
print(ask_posts[:2])
print("\n")
print(show_posts[:2])

[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43']]


[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46']]


In [10]:
def Get_Mean(data, pos):
    sum_col = 0
    for item in data:
        col = int(post[pos]) 
        sum_col += col
        
    return sum_col/len(data);

total_ask_comments = 0

avg_ask_comments = Get_Mean(ask_posts, 4);

print(f'Ask_posts average value: {avg_ask_comments}')

avg_show_comments = Get_Mean(show_posts, 4);

print(f'Show_posts average value: {avg_show_comments}')

Ask_posts average value: 2.0
Show_posts average value: 2.0


They both have same average number of comments.

In [3]:
print(f'Ask_posts records: {len(ask_posts)}')
print(f'Show_posts records: {len(show_posts)}')
print(f'Other_posts records: {len(other_posts)}\n')

print(ask_posts[:2])
print("\n")
print(show_posts[:2])

Ask_posts records: 1744
Show_posts records: 1162
Other_posts records: 17194

[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43']]


[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46']]


## Analysis

Now we will answer some questions to help understand the business and make informed decision.

### Determine Average Comments by Post

let's determine if ask posts or show posts receive more comments on average.

In [4]:
# function to get average date based on specified index

def getAverageByColumn(data, index):
    total_of_column = 0

    for row in data:
        total_of_column += int(row[index])

    avg_of_column = total_of_column/len(data) if total_of_column > 0 else 0
    
    return avg_of_column

In [5]:
avg_ask_comments = getAverageByColumn(ask_posts, 4)
avg_show_comments = getAverageByColumn(show_posts, 4)

print(f'Average Ask_posts comments: {round(avg_ask_comments)}')
print(f'Average show_posts comments: {round(avg_show_comments)}')

Average Ask_posts comments: 14
Average show_posts comments: 10


### Finding
While the computed average above shows that `show_posts` as an average comment of **10**, the average number of comments from ask_posts is higher by 4 (**14**).  


### Finding the Number of Ask Posts and Comments by Hour Created

Next, we'll determine if ask posts created at a certain time are more likely to attract comments. We'll use the following steps to perform this analysis:

- Calculate the number of ask posts created in each hour of the day, along with the number of comments received.
- Calculate the average number of comments ask posts receive by hour created.

In [6]:
import datetime as _dt

In [7]:
result_list = []

# extract date and comments into new list

for row in ask_posts:
    created_at = row[6]
    num_comment = int(row[4])
    result_list.append([created_at,num_comment])
    

# create a frequency table for hour time and comment

counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    dt = row[0]
    dateObj = _dt.datetime.strptime(dt, "%m/%d/%Y %H:%M")
    hr = dateObj.strftime("%H")
    
    if hr not in counts_by_hour:
        counts_by_hour[hr] = 1
        comments_by_hour[hr] = row[1]
    else:
        counts_by_hour[hr] += 1
        comments_by_hour[hr] += row[1]


In [8]:
avg_by_hour = []

# get average number of comment by the hour

for hr in counts_by_hour:
    avg = comments_by_hour[hr]/counts_by_hour[hr]
    avg_by_hour.append([hr, avg])

In [9]:
print(avg_by_hour)

[['09', 5.5777777777777775], ['13', 14.741176470588234], ['10', 13.440677966101696], ['14', 13.233644859813085], ['16', 16.796296296296298], ['23', 7.985294117647059], ['12', 9.41095890410959], ['17', 11.46], ['15', 38.5948275862069], ['21', 16.009174311926607], ['20', 21.525], ['02', 23.810344827586206], ['18', 13.20183486238532], ['03', 7.796296296296297], ['05', 10.08695652173913], ['19', 10.8], ['01', 11.383333333333333], ['22', 6.746478873239437], ['08', 10.25], ['04', 7.170212765957447], ['00', 8.127272727272727], ['06', 9.022727272727273], ['07', 7.852941176470588], ['11', 11.051724137931034]]


In [10]:
# create a swap list with - [average, hour]

swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
    
print(swap_avg_by_hour)

# sort list in descending order

sorted_swap = sorted(swap_avg_by_hour, reverse=True)

print("\n",sorted_swap[:6])

[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]

 [[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21'], [14.741176470588234, '13']]


In [11]:
template = "{hr}:00: {avg:.2f} average comments per post"

for row in sorted_swap[:6]:
    output = template.format(hr=row[1], avg=row[0])
    print(output)

15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post
13:00: 14.74 average comments per post


### Finding

The analysis above shows that 3pm (15:00) has the highest average comments per post, and would be recommended as the best time to create posts because it has a higher chance receiving feedback.


### Determine if show or ask posts receive more points on average.

The posts point represent the variance between number of thumbs up and thumbs down. We will now determine the which of the two posts has the most average points.

In [12]:
avg_ask_points = getAverageByColumn(ask_posts, 3)
avg_show_points = getAverageByColumn(show_posts, 3)

print(f'Average Ask_posts points: {round(avg_ask_points)}')
print(f'Average show_posts points: {round(avg_show_points)}')

Average Ask_posts points: 15
Average show_posts points: 28


**Finding**: show_posts has a higher number of points (28) on posts compared to ask_posts (15).


### Points by Post Time

Next, we would determine if posts created at a certain time are more likely to receive more points.


In [13]:
#override prev. list for mem. optimization

result_list = []

# extract date and comments into new list

for row in hn:
    point = int(row[3])
    post_date = row[6]
    result_list.append([point, post_date])

In [14]:
hn_count_by_hr = {}
hn_point_by_hr = {}

# create a frequency table for hour time and point for hn dataset

for row in result_list:
    dt = row[1]
    dateObj = _dt.datetime.strptime(dt, "%m/%d/%Y %H:%M")
    hr = dateObj.strftime("%H")
    
    if hr not in hn_count_by_hr:
        hn_count_by_hr[hr] = 1
        hn_point_by_hr[hr] = row[0]
    else:
        hn_count_by_hr[hr] += 1
        hn_point_by_hr[hr] += row[0]

In [15]:
hn_avg_by_hr = []

for hr in hn_count_by_hr:
    avg = hn_point_by_hr[hr]/hn_count_by_hr[hr]
    
    # doing the swicth so we don't have to swap
    hn_avg_by_hr.append([avg, hr])
    
# sort list into descending

sorted_hn_avg_by_hr = sorted(hn_avg_by_hr, reverse=True)

In [16]:
template = "{hr}:00: {avg:.2f} average point per post"

for row in sorted_hn_avg_by_hr[:6]:
    output = template.format(hr=row[1], avg=row[0])
    print(output)

13:00: 56.17 average point per post
15:00: 55.65 average point per post
10:00: 54.71 average point per post
14:00: 54.44 average point per post
19:00: 54.17 average point per post
00:00: 53.58 average point per post


### Finding

Based on the analysis above, 1pm (13:00) has the highest average points per post on the overall `Hacker News` posts, and would be recommended as the best time to create posts in order to generate more points.

## Comments and Point By Average on Other Posts

It would also be interesting to understand what the average number of comments and points other posts receive and then compare this with existing findings.

In [17]:
avg_comment_otherposts = getAverageByColumn(other_posts, 4)
avg_point_otherposts = getAverageByColumn(other_posts, 3)

print(f'Average comments on other posts: {round(avg_comment_otherposts)}')
print(f'Average points on other posts: {round(avg_point_otherposts)}')

Average comments on other posts: 27
Average points on other posts: 55


### Finding

On a average, other posts has a higher average for comments and points. On comments, other posts has average of 27 compared to ask_posts (14) and show_posts (10), likewise, other_posts has average of 55 compared to aks_posts (15) and show_posts (28).

## Conclusion

Drawing insights from the report analysis, it is deduced that ask_posts exhibit a greater volume of posts and an elevated average number of comments compared to show_posts. Furthermore, it is noteworthy that the optimal time for post creation is at 3 pm, given its association with the highest average comment count per post within a one-hour timeframe.

Upon conducting further analysis, the figure shows that other_posts demonstrate a superior average in both comments and points when compared with ask_posts and show_posts. This fact can be attributed to the expansive array of topics encompassed within other_posts, which, hypothetically speaking, may stem from yet-to-be-explored subject matter. Notably, other_posts boasts a substantial dataset of `17,194` records, markedly surpassing the respective counts of ask_posts (`1,744`) and show_posts (`1,162`).